teps/0140-producing-results-in-matrix.md

```diff
@@ -0,0 +1,624 @@
+            value:
+              - chrome
+              - safari
+              - firefox
+      taskSpec:
+        params:
+          - name: platform
+            type: string
+          - name: browser
+            type: string
+        results:
+          - name: report-url
+            type: string
+        steps:
+          - name: produce-report-url
+            image: alpine
+            script: |
+              echo "Running tests on $(params.platform)-$(params.browser)"
+              echo -n "https://api.example/get-report/$(params.platform)-$(params.browser)" | tee $(results.report-url.path)
+```
+
+## Proposal
+
+To support enabling a matrixed `pipelineTask` to produce `results`, we propose, each fanned out `taskRun` that
+produces `result` of type `string` will be aggregated into an `array` of `results` during reconciliation, in which the
+whole `array` of `results` can be consumed by another `pipelineTask` using the star notion `[*]`.
+
+| Result Type in `taskRef` or `taskSpec` | Parameter Type of Consumer | Specification                                         |
+|----------------------------------------|----------------------------|-------------------------------------------------------|
+| string                                 | array                      | `$(tasks.<pipelineTaskName>.results.<resultName>[*])` |
+| array                                  | Not Supported              | Not Supported                                         |
+| object                                 | Not Supported              | Not Supported                                         |
+
+## Design Details
+
+### Results Cache
+
+With this proposal, we add a `resultsCache` to `ResolvedPipelineTask` that enables caching of `results` from a
+matrixed `pipelineTask` in order to prevent resolving the result references for the referenced `matrixed pipelineTask`
+on every single reconcile loop.
+
+```go
+t.ResultsCache[result.Name] = []string{result.Value.StringVal}
+```
+
+```go
+type ResolvedPipelineTask struct {
+	TaskRunNames []string
+	TaskRuns     []*v1.TaskRun
+	CustomTask     bool
+	CustomRunNames []string
+	CustomRuns     []*v1beta1.CustomRun
+	PipelineTask   *v1.PipelineTask
+	ResolvedTask   *resources.ResolvedTask
+	ResultsCache   map[string][]string
+}
+```
+
+### Context Variables
+
+We propose enabling `context` variables to allow users to access the `matrix` runtime data.
+
+#### Access Matrix Combinations Length
+
+The pipeline authors can access the total number of instances created as part of the `matrix` using the syntax:
+`tasks.<pipelineTaskName>.matrixLength` and `finally.<pipelineTaskName>.matrixLength`.
+
+#### Access Aggregated Results Length
+
+The pipeline authors can access the length of the array of aggregated results that were
+actually produced using the syntax: `tasks.<pipelineTaskName>.matrix.<resultName>`
+and `finally.<pipelineTaskName>.matrix.<resultName>`. This will allow users to loop over the
```

Comment:
Would `finally.<pipelineTaskName>.matrix.<resultName>` be used in pipeline results? I don't think it could be used in a subsequent task

Reply:
Yes `finally` can be used in the pipeline results.